Add option to skip trusted org members for DCO check

xmudrii opened the pull request:
This PR adds an option to skip the DCO check for the trusted org members and the collaborators. The DCO plugin can now be configured per-repository basis.

xmudrii commented on prow/plugins/config.go:
```diff
@@ -477,6 +478,14 @@ type Welcome struct {
+// Dco is config for the DCO plugin.
+type Dco struct {
+	// Repos is either of the form org/repos or just org.
+	Repos []string `json:"repos,omitempty"`
```
Do I need anything else to make per repo configuration work?

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +137,32 @@ type commentPruner interface {
 }
 
+// checkTrustedOrgMember checks are all commits from the trusted org member
+func checkTrustedOrgMember(dco *plugins.Dco, gc gitHubClient, l *logrus.Entry, org, repo string, number int) (bool, error) {
```
Here we just check for organization members? Should we check also for collaborators? If yes, should that be default or optional?

stevekuznetsov replied:
Look for `prow/plugins/trigger.TrustedUser()`

xmudrii replied:
I have updated the function to match what's done by `TrustedUser()`. Let me know if it's okay.

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +137,32 @@ type commentPruner interface {
+
+	trusted := true
+	for _, commit := range allCommits {
+		member, err := gc.IsMember(dco.TrustedOrg, commit.Author.Login)
```
Is `commit.Author.Login` a correct thing to check here?

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -162,11 +228,38 @@ func checkExistingLabels(gc gitHubClient, l *logrus.Entry, org, repo string, num
 	signedOff := len(commitsMissingDCO) == 0
 
+	if trustedOrgMember {
+		if hasNoLabel {
```
This will remove all DCO labels if the member is trusted? Do we want to remove all DCO labels or set yes DCO label?

stevekuznetsov replied:
@munnerz

xmudrii replied:
I'm not sure is it a good idea to remove both labels because there may be instances that are enforcing the `dco-signoff: yes` label. In that case Prow would not merge the PR. Should we apply the label or require user not to check the label?

munnerz replied:
We personally block tide merges on `dco-signoff: "yes"`, so would want this to at least be possible. I'm not sure if it's worthwhile making it configurable at this point?

munnerz replied:
Perhaps we should consider disabling this requirement and just relying on the GitHub status - I don't think we've considered the options/tradeoffs in much depth

xmudrii replied:
We're doing the same, but I wasn't sure should we keep this behavior or change it. I believe putting it under some option should be easy, so if that makes sense I'll go ahead and add it.

munnerz replied:
+1, I would rather retain the `dco-signoff: "yes"` label for member PRs personally

xmudrii replied:
Fixed, the label should be applied now.

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -162,11 +228,38 @@ func checkExistingLabels(gc gitHubClient, l *logrus.Entry, org, repo string, num
+		}
+		if hasYesLabel {
+			l.Debugf("Removing %q label", dcoYesLabel)
+			// remove 'dco-signoff: no' label
```
Reminder: update comment. :upside_down_face:

stevekuznetsov commented on prow/plugins/dco/dco.go:
```diff
@@ -81,8 +98,31 @@ func helpProvider(config *plugins.Configuration, enabledRepos []string) (*plugin
+}
+
+// strInSlice returns true if any string in slice matches str exactly
+func strInSlice(str string, slice []string) bool {
```
You could just use one loop above and not need to factor this out -- I think that may be more readable, otherwise we would accumulate a lot of these. Alternatively:

```go
for _, repoConfig := range config.Dco {
    if !sets.NewString(repoConfig.Repos...).HasAny(org, fullName) {
        // etc
    }
}
```

xmudrii replied:
I've experimented and used similar snippet in the `DcoFor` function. The unit tests are passing, but let me know if it's okay to use it.

stevekuznetsov commented on prow/plugins/dco/dco.go:
```diff
@@ -81,8 +98,31 @@ func helpProvider(config *plugins.Configuration, enabledRepos []string) (*plugin
+}
+
+// strInSlice returns true if any string in slice matches str exactly
+func strInSlice(str string, slice []string) bool {
```
You could just use one loop above and not need to factor this out -- I think that may be more readable, otherwise we would accumulate a lot of these. Alternatively:

```go
for _, repoConfig := range config.Dco {
    if !sets.NewString(repoConfig.Repos...).HasAny(org, fullName) {
        // etc
    }
}
```

stevekuznetsov commented on prow/plugins/dco/dco.go:
```diff
@@ -162,11 +228,39 @@ func checkExistingLabels(gc gitHubClient, l *logrus.Entry, org, repo string, num
 // takeAction will take appropriate action on the pull request according to its
 // current state.
-func takeAction(gc gitHubClient, cp commentPruner, l *logrus.Entry, org, repo string, pr github.PullRequest, commitsMissingDCO []github.GitCommit, existingStatus string, hasYesLabel, hasNoLabel, addComment bool) error {
+func takeAction(gc gitHubClient, cp commentPruner, l *logrus.Entry, org, repo string, pr github.PullRequest, commitsMissingDCO []github.GitCommit, existingStatus string, hasYesLabel, hasNoLabel, addComment, trustedOrgMember bool) error {
```
Could we not pass around pointers to the config please, and make a lack of configuration equal to the default case?

xmudrii replied:
> Could we not pass around pointers to the config please
Fixed, we're not passing pointers anymore
> make a lack of configuration equal to the default case
What is default case supposed to be here? Do we want default case to be that if config is not provided, always do DCO checks?

stevekuznetsov replied:
The default case should be the same behavior as before this change -- to get the new behavior you must opt-in

xmudrii replied:
Makes sense. To solve this I added the `SkipDcoCheck` field that is by default false but if operator wants to use this feature `SkipDcoChecks` needs to be set to true. Please let me know if you have any better idea.

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -81,8 +98,31 @@ func helpProvider(config *plugins.Configuration, enabledRepos []string) (*plugin
 }
 
+// optionsForRepo gets the plugins.Dco struct that is applicable to the indicated repo.
+func optionsForRepo(config *plugins.Configuration, org, repo string) *plugins.Dco {
```
Yes, I've added a function called `DcoFor` that works similar as `TriggerFor`.

xmudrii commented:
@stevekuznetsov Thanks for the review! I addressed some of your comments and will try to address rest throughout the day. I also left some more questions where I was unsure what to do.

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -66,10 +66,27 @@ func init() {
+	for _, orgRepo := range enabledRepos {
+		parts := strings.Split(orgRepo, "/")
+		var opts *plugins.Dco
+		switch len(parts) {
```
I'm not sure I understand this comment. Should I simplify this for the DCO plugin or simplify this everywhere?

If it's the latter an option could be to have a global helper that would take `orgRepo` and return org and repo, so we don't have to do switch everywhere:

```go
// If just org is passed to as the argument return "" for repo
org, repo := plugins.SplitOrgRepo(orgRepo)
opts := config.DcoFor(org, repo)
```

fejta commented on prow/plugins/config.go:
```diff
@@ -477,6 +478,20 @@ type Welcome struct {
 	MessageTemplate string `json:"message_template,omitempty"`
 }
 
+// Dco is config for the DCO plugin.
```
Please define DCO somewhere

xmudrii replied:
I added a link there. It is also defined like that in the `dco` package.

xmudrii commented on prow/plugins/config.go:
```diff
@@ -60,6 +60,7 @@ type Configuration struct {
 	Cat                        Cat                    `json:"cat,omitempty"`
 	CherryPickUnapproved       CherryPickUnapproved   `json:"cherry_pick_unapproved,omitempty"`
 	ConfigUpdater              ConfigUpdater          `json:"config_updater,omitempty"`
+	Dco                        []Dco                  `json:"dco,omitempty"`
```
I've tried to get around this, but I'm not sure how this should look like and where exactly it's implemented for `branch_protection`. Can you give me some hints or examples where can I find that or how that should look like? I've went through #9757 but it's mostly about implementing a library that does this and there's still some ongoing discussion.

stevekuznetsov replied:
Example is here: https://github.com/kubernetes/test-infra/blob/master/prow/config/branch_protection.go

stevekuznetsov replied:
Basic idea is having a hierarchy of embedded structs

alvaroaleman replied:
Look here for a sample: https://github.com/kubernetes/test-infra/blob/0e68a1e45538e8066910c92552df926418df0020/prow/config/config.go#L219-L233

Can also replace the logic in the plugins help, IMHO much easier to understand than the current impementation.

alvaroaleman replied:
(Use a `map[string]*DCO`)

xmudrii replied:
I've followed the example Alvaro linked to try to implement hierarchies and overrides. The way Plank does it is a little bit different compared to `branch_protection` but on the other side it seems more simple. If I implemented it correctly, I believe that we will get the effect we want and that it's going to be easier to manage DCO settings per org and repo.

The way it works is similar to Plank. The Dco field is now `map[string]*Dco` and the key scheme is:
* `*` is the default DCO configuration (don't skip DCO checks),
* if `org/repo` key exists, return it
* if `org` key exists, return it
* otherwise return default (`*`)

A potential problem here is to make this work, `Dco` has to be `map[string]*Dco`. Earlier, there have been some concerns regarding using pointers, but I think using pointers is the only way to make it work. If there is a better way, please let me know.

@alvaroaleman @stevekuznetsov PTAL and let me know what do you think and is there anything I can improve.

xmudrii commented:
I've went through all comments and tried to fix most of them. The WIP tag is removed from the PR and the PR is ready for review.

I've tried to add many unit tests for this behavior and by unit tests it seems like this works as expected. I've not tested this on a real Prow instance as I'm not sure is this possible at all. I've tried to make this with backwards-compatibility in mind, so behavior should stay the same if the plugin isn't configured.

I wasn't sure how to address all comments, but the most important ones should be addressed. Here are some comments that I wasn't sure how to address:

* Do we want to apply the `dco-signoff: yes` label or not apply any label at all? https://github.com/kubernetes/test-infra/pull/11832#discussion_r266841104
* How the `Dco` structure should look like? The idea was to make it something like `branch_protection` (#9757) but I was not sure how exactly it should look like. https://github.com/kubernetes/test-infra/pull/11832#discussion_r267024216
* There was a request to DRY the pluginHelp little bit, but I wasn't sure should I do on the plugin-level or make a structure that would work for all plugins. https://github.com/kubernetes/test-infra/pull/11832#discussion_r267024400

/cc @stevekuznetsov @munnerz @fejta

stevekuznetsov commented on prow/plugins/config.go:
```diff
@@ -612,6 +627,18 @@ func (c *Configuration) TriggerFor(org, repo string) Trigger {
+// DcoFor finds the Dco for a repo, if one exists
+// a Dco can be listed for the repo itself or for the
+// owning organization
+func (c *Configuration) DcoFor(org, repo string) Dco {
```
This approach does not really allow for repo-level overrides of org-level fields. I think I linked to #9757 before -- do we care?

stevekuznetsov replied:
(you can get org-level overrides or repo-level overrides with appropriate ordering, but that's not obvious)

xmudrii replied:
@stevekuznetsov I've looked into this and I think I better understand how it works. If this is something we want, I can give it a try. Do we want to give it a try or wait for a library to be implemented?

alvaroaleman replied:
>  or wait for a library to be implemented?
We need generics for that :S

stevekuznetsov replied:
There's an attempt to codegen it currently WIPO

stevekuznetsov commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +116,54 @@ type commentPruner interface {
 }
 
+// checkTrustedUser checks are all commits from a trusted user
+func checkTrustedUser(dco plugins.Dco, gc gitHubClient, l *logrus.Entry, org, repo string, number int) (bool, error) {
```
Why are we not using the logic from `trigger`?

xmudrii replied:
Fixed. However, I wasn't sure where I should put the `TrustedUser` function. Currently, it's in `prow/plugins/trusteduser.go`, but if there's a better place, let me know. :)

fejta commented:
I'm going to delegate this review to Steve as I am not familiar with the needs for a DCO process.
/uncc

stevekuznetsov commented:
I'm also just doing code hygiene reviews, @munnerz would love your overall review here :)

xmudrii commented:
@munnerz @stevekuznetsov Sorry for such a long delay, it has been very busy over here! I've fixed the comments @stevekuznetsov left earlier. Please take a look and let me know if it's okay now. :slightly_smiling_face:

munnerz commented on prow/plugins/dco/dco_test.go:
```diff
@@ -225,6 +234,205 @@ Instructions for interacting with me using PR comments are available [here](http
+					SHA:    "sha",
+					Commit: github.GitCommit{Message: "not signed off"},
+					Author: github.User{
+						Login: "test-2",
```
Should the Login here be `test` (i.e. an org member) as we are validating that `SkipDCOCheck: false` with `TrustedOrg` set causes the checker to always fail if commits aren't signed off?

xmudrii replied:
Yes, I believe it should be `test` instead.

munnerz commented on prow/plugins/dco/dco_test.go:
```diff
@@ -381,6 +596,35 @@ Instructions for interacting with me using PR comments are available [here](http
 `,
 		},
+		{
+			name: "should success as skip dco is enabled",
```
nit: succeed

munnerz commented on prow/plugins/trigger/generic-comment.go:
```diff
@@ -18,9 +18,10 @@ package trigger
 	"regexp"
 
+	"k8s.io/test-infra/prow/pjutil"
+
```
nit: remove newline

munnerz commented on prow/plugins/trusteduser.go:
```diff
@@ -0,0 +1,59 @@
+limitations under the License.
+*/
+
+package plugins
```
I'm not sure where the right place for this method is - @stevekuznetsov any suggestions/thoughts?

xmudrii replied:
I was not really sure where to put it as well. Any pointers would be appreciated. :grimacing:

stevekuznetsov replied:
Not sure it needed to move?

munnerz commented on prow/plugins/trusteduser_test.go:
```diff
@@ -0,0 +1,99 @@
+	}{
+		{
+			name:            "user is member of trusted org",
+			onlyOrgMembers:  false,
```
nit: maybe add a test case that has `onlyOrgMembers: true` & `user: "test"`, just for the extra coverage?

xmudrii replied:
Done. :+1:

xmudrii commented on prow/plugins/config.go:
```diff
@@ -480,6 +481,20 @@ type Welcome struct {
+	// Repos is either of the form org/repos or just org.
+	Repos []string `json:"repos,omitempty"`
+	// SkipDcoCheck is used to skip DCO check for collaborators and org members
+	SkipDcoCheck bool `json:"skip_dco_check,omitempty"`
```
I'm not really sure, but I'd be happy to change it if needed. :slightly_smiling_face:

xmudrii replied:
Fixed

xmudrii commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +116,32 @@ type commentPruner interface {
+	}
+
+	for _, commit := range allCommits {
+		trusted, err := plugins.TrustedUser(gc, dco.OnlyOrgMembers, dco.TrustedOrg, commit.Author.Login, org, repo)
```
As far as I understand, yes, that's the user that pushed the commit. The idea was to match the current behavior as close as possible—if non-org member pushes a commit and it's not signed, it'll fail regardless of is trusted users that pushed that commit.

xmudrii commented:
@munnerz @stevekuznetsov Sorry for such a long delay yet again! I went through comments and addressed most of them. There are some outstanding questions and concerns, so I'd like some input on those. :slightly_smiling_face: The PR is also rebased.

nikhita commented:
:wave: bump, looks like this is ready for a review :)

stevekuznetsov commented on prow/plugins/trigger/generic-comment.go:
```diff
@@ -74,7 +74,7 @@ func handleGenericComment(c Client, trigger plugins.Trigger, gc github.GenericCo
 
 	// Skip untrusted users comments.
-	trusted, err := TrustedUser(c.GitHubClient, trigger, commentAuthor, org, repo)
+	trusted, err := plugins.TrustedUser(c.GitHubClient, trigger.OnlyOrgMembers, trigger.TrustedOrg, commentAuthor, org, repo)
```
This was already exported, why did it need to move?

xmudrii replied:
@stevekuznetsov The reason why I moved it is because it was in the `trigger` package before. As it's used by both `trigger` and `dco` packages now, I thought it should be placed at the higher level. I'm open to suggestions should I move it back or move to some other place.

stevekuznetsov replied:
I think we can just keep it in place

/shrug

xmudrii replied:
@stevekuznetsov Just to make sure I understood correctly. Keep it in the place as in return it to the `trigger` package or keep it in the current place (`prow/plugins/trusteduser.go`)? :sweat_smile:

stevekuznetsov replied:
Keep it in `trigger` -- will make these changes smaller

xmudrii replied:
I moved it back to `trigger`. However, I kept the change which made it non-dependent on the `Trigger` structure, which required me to do this https://github.com/kubernetes/test-infra/pull/11832/commits/bd45e88813a015b2cefc7e98f341738341d61082? WDYT about this?

nikhita commented on prow/plugins/config.go:
```diff
@@ -61,6 +61,7 @@ type Configuration struct {
 	Cat                        Cat                    `json:"cat,omitempty"`
 	CherryPickUnapproved       CherryPickUnapproved   `json:"cherry_pick_unapproved,omitempty"`
 	ConfigUpdater              ConfigUpdater          `json:"config_updater,omitempty"`
+	Dcos                       []Dco                  `json:"dco,omitempty"`
```
Can we keep this as `Dco` instead of `Dcos` to be consistent with the json `dco`?

xmudrii replied:
Fixed

nikhita commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +117,32 @@ type commentPruner interface {
+			return false, fmt.Errorf("Error checking is member trusted: %v", err)
+		}
+		if !trusted {
+			l.Debugf("Member %s is neither org member or repo collaborator", commit.Author.Login)
```
Can we just say that the member is not trusted here? If `onlyOrgMembers` is true, I'm not sure we can accurately say if the commit author is not a repo collaborator...

xmudrii replied:
Fixed

nikhita commented on prow/plugins/dco/dco.go:
```diff
@@ -97,6 +117,32 @@ type commentPruner interface {
+// checkTrustedUser checks are all commits from a trusted user
+func checkTrustedUser(dco plugins.Dco, gc gitHubClient, l *logrus.Entry, org, repo string, number int) (bool, error) {
+	// Skip checking is user trusted if this option isn't enabled.
+	if !dco.SkipDcoCheck {
```
nit: if you move this outside this function (in `handle`), you can avoid passing the dco config and reduce a function call

xmudrii replied:
Fixed

xmudrii commented on prow/BUILD.bazel:
```diff
@@ -130,6 +130,7 @@ filegroup(
         "//prow/labels:all-srcs",
         "//prow/logrusutil:all-srcs",
         "//prow/metrics:all-srcs",
+        "//prow/monitoring:all-srcs",
```
@stevekuznetsov @munnerz Any idea why this got added here and why `prow/monitoring/BUILD.bazel` got changed? I haven't touched anything related at all.

stevekuznetsov replied:
Yeah, there was an error -- rebasing on `master` will fix this

munnerz commented on prow/plugins/dco/dco_test.go:
```diff
@@ -225,6 +234,205 @@ Instructions for interacting with me using PR comments are available [here](http
 			expectedStatus: github.StatusSuccess,
 		},
+		{
+			name: "should skip dco check as user is member of the trusted org (all commits signed)",
```
should be "all commits non-signed"? or the `commit` needs to be updated otherwise 😄

xmudrii replied:
Fixed and updated tests to reflect changes related to labels. PTAL

alvaroaleman commented on prow/plugins/config.go:
```diff
@@ -466,6 +467,21 @@ type Welcome struct {
+	// OnlyOrgMembers requires PRs from non-org members to have a valid DCO signoff message
+	// if the skip DCO option is enabled.
+	// By default, SkipDCOCheck also includes repo collaborators.
+	OnlyOrgMembers bool `json:"only_org_members,omitempty"`
```
Instead of having one bool that skips for both collaborator/member and then another bool that if the first bool is enabled re-enables not skipping for collaborators - Can we just have one bool `SkipDCOCheckForCollaborators` and another `SkipDCOCheckForMembers`?

alvaroaleman commented on prow/plugins/config.go:
```diff
@@ -621,6 +633,22 @@ func (c *Configuration) TriggerFor(org, repo string) Trigger {
+// a Dco can be listed for the repo itself or for the
+// owning organization
+func (c *Configuration) DcoFor(org, repo string) *Dco {
+	if org == "" && repo == "" {
```
I don't understand this case, when does this get called with empty `org` and `repo`?

xmudrii replied:
Fixed

alvaroaleman commented on prow/plugins/config.go:
```diff
@@ -621,6 +633,22 @@ func (c *Configuration) TriggerFor(org, repo string) Trigger {
+	if c.Dco[org] != nil {
+		return c.Dco[org]
+	}
+	return c.Dco["*"]
```
Can youn add a check if `c.DCO["*"]` is nil and if yes, return a `&DCO{}`? Consumers of this func will not expect to get a nilpointer back

xmudrii replied:
Added

alvaroaleman commented on prow/plugins/dco/dco.go:
```diff
@@ -66,10 +67,27 @@ func init() {
+	for _, orgRepo := range enabledRepos {
+		parts := strings.Split(orgRepo, "/")
+		var opts *plugins.Dco
+		switch len(parts) {
```
@stevekuznetsov Do we have some helper that takes an orgRepo string and returns two strings, org and repo? I imagine this to be needed in many places

alvaroaleman commented on prow/plugins/dco/dco.go:
```diff
@@ -66,10 +67,27 @@ func init() {
+			opts = config.DcoFor(parts[0], "")
+		case 2:
+			opts = config.DcoFor(parts[0], parts[1])
+		default:
```
Just return an error? It doesnt make sense to have an empty string in `enabledRepos`

xmudrii replied:
Fixed

xmudrii commented:
@alvaroaleman Addressed comments you left, PTAL

xmudrii commented:
/unshrug

alvaroaleman commented on prow/plugins/config.go:
```diff
@@ -743,6 +771,15 @@ See also [Kubernetes Patch Releases](https://github.com/kubernetes/sig-release/b
 		}
 	}
+
+	if c.Dco == nil {
```
This is not needed

xmudrii replied:
Removed, this was a leftover.

xmudrii commented:
Addressed all comments so far, PTAL

alvaroaleman commented:
/lgtm

munnerz commented:
I'm not named as an OWNER, so this will require someone else to `/approve` 😄 

/lgtm

xmudrii commented:
No changes, just rebased and squashed

munnerz commented:
Assigning Steve for approval 😄 

/assign @stevekuznetsov

munnerz commented:
/lgtm
